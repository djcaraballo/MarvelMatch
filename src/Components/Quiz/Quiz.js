import React, { Component } from 'react'
import PropTypes from 'prop-types'
import QuizContainer from '../QuizContainer/QuizContainer'
import Question from '../Questions/Questions'
import quizQuestions from '../../Utils/API/quizQuestions'
import Result from '../Result/Result'

class Quiz extends Component {
  constructor (props) {
    super(props)
    this.state = {
      counter: 0,
      questionId: 1,
      question: '',
      answerOptions: [],
      answer: '',
      answersCount: {
        blueEyedHeroes: 0,
        brownEyedHeroes: 0,
        greenEyedHeroes: 0,
        otherColoredEyedHeroes: 0,
        blondeHairedHeroes: 0,
        blackHairedHeroes: 0,
        redHairedHeroes: 0,
        brownHairedHeroes: 0,
        otherHairedHeroes: 0,
        tallHeroes: 0,
        shortHeroes: 0,
        averageHeroes: 0,
      },
      result: ''
      //change to object once everything is working 
    }
  }

  componentWillMount() {
    this.setState({
      question: quizQuestions[0].question,
      answerOptions: quizQuestions[0].answers
    })
  }

  handleAnswerSelected = (e) => {
    this.setUserAnswer(e.currentTarget.value)
    if(this.state.questionId < quizQuestions.length) {
      setTimeout(() => this.setNextQuestion(), 300)
    } else {
      setTimeout(() => this.setResults(this.getResults()), 300)
    }
  }

  setUserAnswer(answer) {
    this.setState(state => 
      ({
      answersCount: {
        ...state.answersCount,
        [answer]: state.answersCount[answer]++
      },
      answer: answer
    }))
  }

  setNextQuestion() {
    const counter = this.state.counter + 1
    const questionId = this.state.questionId + 1
    this.setState({
      counter: counter,
      questionId: questionId,
      question: quizQuestions[counter].question,
      answerOptions: quizQuestions[counter].answers,
      answer: '',
    })
  }

  getResults() {
    const answersCount = this.state.answersCount
    const answersCountKeys = Object.keys(answersCount)
    const answersCountValues = answersCountKeys.map((key) => answersCount[key])
    const maxAnswerCount = Math.max.apply(null, answersCountValues)
    return answersCountKeys.filter((key) => answersCount[key] === maxAnswerCount)
  }

  setResults(result) {
    if (result.length === 1) {
      this.setState({ result: result[0] })
    } else {
      this.setState({ result: 'Undetermined' })
    }
  }

  renderQuizContainer() {
    const { answer, answerOptions, questionId, question } = this.state
    return (
      <QuizContainer 
        answer={answer}
        answerOptions={answerOptions}
        questionId={questionId}
        question={question}
        questionTotal={quizQuestions.length}
        onAnswerSelected={this.handleAnswerSelected}
      />
    )
  }

  renderResult() {
    return (
      <Result quizResult={this.state.result} />
    )
  }

  render () {
    return (
      <div className="quiz-wrapper" key={this.state.questionId}>
        <h1>Take the Quiz and match a Marvel Character!</h1>
        {this.state.result ? this.renderResult() : this.renderQuizContainer()}
      </div>
    )
  }
}

export default Quiz
