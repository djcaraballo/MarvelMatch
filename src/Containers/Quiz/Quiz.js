import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import QuizContainer from '../../Components/QuizContainer/QuizContainer'
import quizQuestions from '../../Utils/API/quizQuestions'
import Result from '../../Components/Result/Result'

export class Quiz extends Component {
  constructor (props) {
    super(props)
    this.state = {
      counter: 0,
      questionId: 1,
      question: '',
      answerOptions: [],
      answer: '',
      answersCount: {
        blue: 0,
        brown: 0,
        green: 0,
        other: 0,
        blond: 0,
        black: 0,
        red: 0,
        auburn: 0,
        alternate: 0,
        human: 0,
        alien: 0,
        spirit: 0,
        mutant: 0,
        cyborg: 0
      },
      result: ''
    }
  }

  componentWillMount() {
    this.setState({
      question: quizQuestions[0].question,
      answerOptions: quizQuestions[0].answers
    })
  }

  handleAnswerSelected = (e) => {
    this.setUserAnswer(e.target.value)
    if(this.state.questionId < quizQuestions.length) {
      setTimeout(() => this.setNextQuestion(), 400)
    } else {
      setTimeout(() => this.setResults(this.getResults()), 400)
    }
  }

  setUserAnswer(answer) {
    this.setState({
      answersCount: {
        ...this.state.answersCount, 
        [answer]: this.state.answersCount[answer]+=1
      },
      answer: answer
    })
  }

  setNextQuestion() {
    const counter = this.state.counter + 1
    const questionId = this.state.questionId + 1
    this.setState({
      counter: counter,
      questionId: questionId,
      question: quizQuestions[counter].question,
      answerOptions: quizQuestions[counter].answers,
      answer: '',
    })
  }

  getResults() {
    const answersCount = this.state.answersCount
    const answersCountKeys = Object.keys(answersCount)
    const answersCountValues = answersCountKeys.map(key => answersCount[key])
    const possibleResults = this.props.characters.filter((character) => {
      answersCountKeys.forEach((key) => {
        if(character.appearance.hairColor && key === character.appearance.hairColor.toLowerCase() && answersCount[key] !== 0) {
          character.matchTally++
        } else if(character.appearance.eyeColor && key === character.appearance.eyeColor.toLowerCase() && answersCount[key] !== 0) {
          character.matchTally++
        } else if(character.appearance.race && key === character.appearance.race.toLowerCase() && answersCount[key] !== 0) {
          character.matchTally++
        }
      })
      return character.matchTally !== 0
    })
    return possibleResults
  }

  setResults(result) {
    const index = Math.floor(Math.random() * Math.floor(result.length))
    if (result.length >= 1) {
      this.setState({ result: result[index].name })
    } else {
      this.setState({ result: 'Undetermined' })
    }
  }

  handleCharacterMatch = () => {
    const charMatch = this.props.characters.find((character) => {
      return character.name === this.state.result
    })
    return charMatch
  }

  renderQuizContainer() {
    const { answer, answerOptions, questionId, question } = this.state
    return (
      <QuizContainer 
        answer={answer}
        answerOptions={answerOptions}
        questionId={questionId}
        question={question}
        questionTotal={quizQuestions.length}
        onAnswerSelected={this.handleAnswerSelected}
      />
    )
  }

  renderResult() {
    const charMatch = this.handleCharacterMatch()
    return (
      <Result 
        quizResult={this.state.result}
        characterMatch={charMatch} 
      />
    )
  }

  render () {
    return (
      <div className="quiz-wrapper" key={this.state.questionId}>
        <h1>Take the Quiz and match a Marvel Character!</h1>
        {this.state.result ? this.renderResult() : this.renderQuizContainer()}
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  characters: state.characters
})

Quiz.propTypes = {
  characters: PropTypes.array
}

export default connect(mapStateToProps, null)(Quiz)
