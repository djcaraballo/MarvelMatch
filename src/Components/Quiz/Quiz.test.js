import React from 'react'
import Quiz from './Quiz'
import { shallow } from 'enzyme'

describe('Quiz Component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <Quiz />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  xit('should start with the correct default state', () => {

  })

  xit('should set state with a question and an array of answerOptions', () => {

  })

  describe('setUserAnswer function', () => {
    xit('should set state with the correct values for answersCount and answer', () => {

    })
  })

  describe('setNextQuestion function', () => {
    xit('should set state with the correct values', () => {

    })
  })

  describe('setResults function', () => {
    xit('should set state with the correct result if there is a result', () => {

    })

    xit('should set state with Undetermined if there is no result', () => {

    })
  })

  describe('renderQuizContainer function', () => {
    xit('should return an instance of the QuizContainer component', () => {

    })
  })

  describe('renderResult function', () => {
    xit('should return an instance of the Result component', () => {

    })
  })
})