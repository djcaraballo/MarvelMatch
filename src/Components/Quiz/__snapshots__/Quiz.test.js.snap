// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Quiz />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Take the Quiz and match a Marvel Character!
        </h1>,
        <QuizContainer
          answer=""
          answerOptions={
            Array [
              Object {
                "content": "Blue",
                "type": "blueEyedHeroes",
              },
              Object {
                "content": "Brown",
                "type": "brownEyedHeroes",
              },
              Object {
                "content": "Green",
                "type": "greenEyedHeroes",
              },
              Object {
                "content": "Other",
                "type": "otherColoredEyedHeroes",
              },
            ]
          }
          onAnswerSelected={[Function]}
          question="What color are your eyes?"
          questionId={1}
          questionTotal={3}
        />,
      ],
      "className": "quiz-wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Take the Quiz and match a Marvel Character!",
        },
        "ref": null,
        "rendered": "Take the Quiz and match a Marvel Character!",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "answer": "",
          "answerOptions": Array [
            Object {
              "content": "Blue",
              "type": "blueEyedHeroes",
            },
            Object {
              "content": "Brown",
              "type": "brownEyedHeroes",
            },
            Object {
              "content": "Green",
              "type": "greenEyedHeroes",
            },
            Object {
              "content": "Other",
              "type": "otherColoredEyedHeroes",
            },
          ],
          "onAnswerSelected": [Function],
          "question": "What color are your eyes?",
          "questionId": 1,
          "questionTotal": 3,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Take the Quiz and match a Marvel Character!
          </h1>,
          <QuizContainer
            answer=""
            answerOptions={
              Array [
                Object {
                  "content": "Blue",
                  "type": "blueEyedHeroes",
                },
                Object {
                  "content": "Brown",
                  "type": "brownEyedHeroes",
                },
                Object {
                  "content": "Green",
                  "type": "greenEyedHeroes",
                },
                Object {
                  "content": "Other",
                  "type": "otherColoredEyedHeroes",
                },
              ]
            }
            onAnswerSelected={[Function]}
            question="What color are your eyes?"
            questionId={1}
            questionTotal={3}
          />,
        ],
        "className": "quiz-wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Take the Quiz and match a Marvel Character!",
          },
          "ref": null,
          "rendered": "Take the Quiz and match a Marvel Character!",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "answer": "",
            "answerOptions": Array [
              Object {
                "content": "Blue",
                "type": "blueEyedHeroes",
              },
              Object {
                "content": "Brown",
                "type": "brownEyedHeroes",
              },
              Object {
                "content": "Green",
                "type": "greenEyedHeroes",
              },
              Object {
                "content": "Other",
                "type": "otherColoredEyedHeroes",
              },
            ],
            "onAnswerSelected": [Function],
            "question": "What color are your eyes?",
            "questionId": 1,
            "questionTotal": 3,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
